EIPS/eip-7917.md

```diff
@@ -0,0 +1,103 @@
+
+## Specification
+
+The `BeaconState` container is extended with a `proposer_lookahead` field, which is a list of validator indices covering the full visible lookahead period, starting from the beginning of the current epoch to the next `MIN_SEED_LOOKAHEAD` epochs.
+
+```python
+class BeaconState:
+    ...
+    proposer_lookahead: List[ValidatorIndex, (MIN_SEED_LOOKAHEAD + 1) * SLOTS_PER_EPOCH]
+```
+
+For example, `proposer_lookahead[0]` is the validator index for the first proposer in the current epoch, `proposer_lookahead[SLOTS_PER_EPOCH + 4]` is the validator index for the fifth proposer in the next epoch, and so forth.
+
+The function `get_beacon_proposer_index` is modified to use the pre-calculated `proposer_lookahead` instead of calculating proposer indices on-demand.
+
+```python
+def get_beacon_proposer_index(state: BeaconState) -> ValidatorIndex:
+    """
+    Return the beacon proposer index at the current slot.
+    """
+    return state.proposer_lookahead[state.slot % SLOTS_PER_EPOCH]
+```
+
+At the epoch boundary, the `proposer_lookahead` is updated by shifting out the current epoch’s lookahead and appending the new one.
+
+```python
+def process_epoch(state: BeaconState) -> None:
+    ...
+    process_proposer_lookahead(state)
+
+def process_proposer_lookahead(state: BeaconState) -> None:
+    last_epoch_start = len(state.proposer_lookahead) - SLOTS_PER_EPOCH
+    # Shift out proposers in the first epoch
+    state.proposer_lookahead[:last_epoch_start] = state.proposer_lookahead[SLOTS_PER_EPOCH:]
+    # Fill in the last epoch with new proposer indices
+    last_epoch_proposers = compute_proposer_indices(state, Epoch(get_current_epoch(state) + MIN_SEED_LOOKAHEAD))
+    state.proposer_lookahead[last_epoch_start:] = last_epoch_proposers
+```
+
+See the [updated beacon chain specs](https://github.com/ethereum/consensus-specs/blob/6a1d923730c0b368b5192e64f643f22c21451058/specs/fulu/beacon-chain.md) in consensus-specs for more details.
+
+## Rationale
+
+An alternative approach would be to cache the effective balances at the start of epoch `N` so it can be used to calculate the proposer lookahead at the start of epoch `N+MIN_SEED_LOOKAHEAD`. However, this approach would require additional footprint in the beacon state, and would not be able to provide the proposer lookahead to the EVM via the beacon root.
+
+## Backwards Compatibility
+
+No backward compatibility issues found.
+
+## Test Cases
+
+See [the spec tests](https://github.com/ethereum/consensus-specs/tree/6a1d923730c0b368b5192e64f643f22c21451058/tests/core/pyspec/eth2spec/test/fulu) in consensus-specs.
+
+## Security Considerations
+
+### First Block After the Fork
+
+The first block after the fork will calculate all lookaheads for epochs up to `MIN_SEED_LOOKAHEAD` ahead and fill the `proposer_lookahead` field in the beacon state.
```

Comment:
why is it a security consideration, may be move this to specification section

Reply:
Agree with all the comments around not being a security consideration, let me try to restructure and move things around.